instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/internal/ServerContexts.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.netty.v4_1.internal;
+
+import io.netty.channel.Channel;
+import io.netty.util.Attribute;
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * A helper class for keeping track of incoming requests and spans associated with them.
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+public final class ServerContexts {
```

Comment:
It looks like this class is not thread safe. Is it assumed/known that an instance is only ever used on a single thread or ?

Reply:
This queue is used to support http pipelining (before we just kept a single context in the attribute). Http pipelining requires the responses to be sent in the exact order as requests were received. Due to this I think we can assume that there must be some external coordination going on to maintain this ordering. Anyway since http pipelining isn't the common case mostly this queue will hold a single element.